spec/system/discourse_connect_spec.rb

```diff
@@ -13,24 +13,106 @@
   fab!(:private_topic) { Fabricate(:topic, category: private_category) }
   fab!(:private_post) { Fabricate(:post, topic: private_topic) }
 
+  fab!(:topic)
+  fab!(:post) { Fabricate(:post, topic:) }
+
   before do
     setup_test_sso_server
     configure_discourse_connect
   end
 
   after { shutdown_test_sso_server }
 
-  context "when auth_immediately is enabled" do
-    before { SiteSetting.auth_immediately = true }
+  shared_examples "redirects to SSO" do
+    it "redirects to SSO" do
+      wait_for { has_css?("#current-user") }
+      expect(page).to have_css("a[data-topic-id='#{private_topic.id}']")
+    end
+  end
 
-    it "redirects the user back to the landing URL" do
-      visit private_topic.url
+  shared_examples "shows the homepage" do
+    it "shows the homepage" do
+      expect(page).to have_css("a[data-topic-id='#{topic.id}']")
+    end
+  end
 
-      find(".login-button").click
+  shared_examples "shows the login splash" do
+    it "shows the login splash" do
+      expect(page).to have_css(".login-page")
+    end
+  end
 
-      wait_for { has_css?("#current-user") }
+  context "when login_required is false" do
+    before { SiteSetting.login_required = false }
+
+    context "when auth_immediately is false" do
+      before { SiteSetting.auth_immediately = false }
+
+      context "when visiting /" do
+        before { visit "/" }
+        it_behaves_like "shows the homepage"
+      end
+
+      context "when visiting /login" do
+        before { visit "/login" }
+        it_behaves_like "redirects to SSO"
+      end
+    end
+
+    context "when auth_immediately is true" do
+      before { SiteSetting.auth_immediately = true }
+
+      context "when visiting /" do
+        before { visit "/" }
+        it_behaves_like "shows the homepage"
+      end
+
+      context "when visiting /login" do
+        before { visit "/login" }
+        it_behaves_like "redirects to SSO"
+      end
+
+      it "redirects the user back to the landing URL" do
+        visit private_topic.url
+
+        find(".login-button").click
+
+        wait_for { has_css?("#current-user") }
+
+        expect(page).to have_current_path(private_topic.relative_url)
+      end
+    end
+  end
+
+  context "when login_required is true" do
+    before { SiteSetting.login_required = true }
+
+    context "when auth_immediately is false" do
+      before { SiteSetting.auth_immediately = false }
+
+      context "when visiting /" do
+        before { visit "/" }
+        it_behaves_like "shows the login splash"
+      end
+
+      context "when visiting /login" do
+        before { visit "/login" }
+        it_behaves_like "shows the login splash"
```

Comment:
I pulled the PR locally, and there's a problem with this use case. When I go to `/` and click "Log In" the route changes, but the auth redirect doesn't happen. 

I don't think we need this complexity to be honest, I think in this case we should redirect to SSO when user goes to `/login` directly, regardless of `auth_immediately`.